source/main/physics/BeamFactory.cpp

```diff
@@ -95,14 +95,17 @@ ActorManager::ActorManager()
     , m_simulation_speed(1.0f)
     , m_actor_counter(0)
 {
-    // Create thread pool
+    // Create worker thread (used for physics calculations)
+    m_sim_thread_pool = std::unique_ptr<ThreadPool>(new ThreadPool(1));

```

Comment:
?? I'm guessing you removed the old `ThreadPool(1)` in some recent commit and now you're re-introducing it, otherwise I don't understand this.

Reply:
We removed it a few months ago during the major multi-threading cleanup: https://github.com/RigsOfRods/rigs-of-rods/pull/1598

But it turned out that it can deadlock (as I always suspected): https://github.com/only-a-ptr/rigs-of-rods/pull/22#issuecomment-423998730

See: https://github.com/ulteq/rigs-of-rods/commit/256b8813d3c2f21ad4b2880addae4de36a2617d1#diff-e11c82258dd7c1910ddc4dbc51cafa60

Reply:
`m_sim_thread_pool` is just a lazy way to launch our physics thread in the background.

The alternative would be something like `std::async`.

Reply:
@ulteq Please visit https://github.com/RigsOfRods/rigs-of-rods/issues/816 and explain why it should deadlock the way you describe. I can't think of any reason, other than flawed architecture. I know at the beginning (v0.4.0x) you had to work your way thru all the legacy logic, but I think now we can just do it right without compromises.

> https://github.com/ulteq/rigs-of-rods/commit/256b8813d3c2f21ad4b2880addae4de36a2617d1#diff-e11c82258dd7c1910ddc4dbc51cafa60

🤢

Reply:
The single thread pool approach fails if there is just one worker, because there is no other thread to trigger the condition variable again: https://github.com/RigsOfRods/rigs-of-rods/blob/master/source/main/threadpool/ThreadPool.h#L63, if there is already a waiting task (in our case the sim thread task, which is also waiting at this point).

The alternative solution is to not allow the creation of thread pools with less than 2 workers: https://github.com/RigsOfRods/rigs-of-rods/commit/256b8813d3c2f21ad4b2880addae4de36a2617d1